### ISSUE TITLE
Reader#restoreLines(lines) causes Drop-Line on include::{var}file.ad[]

### ISSUE THREAD

**Issue by mmews-n4** [asciidoctor/asciidoctor#1962]
The problem is triggered only when using a `Preprocessor`, and calling `restoreLines(List<String>)` in the process method: Any include directive that contains a variable (e.g.: `include::{var}file.ad[]` is omitted. 

The command line says:
`WARNING: dropping line containing reference to missing attribute: var`

Example adoc file (assuming an `otherFile.adoc` exists):
```
= My Title
:var: dir/
include::{var}otherFile.adoc[]
End of file.
```

The triggering source line for this behaviour is (reader.rb line 813):
`if (target = @document.sub_attributes raw_target, :attribute_missing => 'drop-line').empty?`

This behaviour is odd since the variable can be resolved when no preprocessor calls the method `restoreLines`.

**Comment by mojavelinux** [comment]
This is actually how this method is designed. The `restore_lines` method (mapped as `restoreLines` in Java) assumes that the content is already preprocessed. No additional preprocessing is done once you put lines onto the reader this way.

There are three ways to add (or replace) lines in a preprocessor so that preprocessing is still performed. I'll show the examples in Ruby.

Let's assume we've defined the following variable containing the content to restore:

```ruby
replacement_content = %(= Document Title
:var: dir/

include::{var}other-file.adoc[]

End of file.)
```
1. You can return a new PreprocessorReader:

```ruby
Asciidoctor::PreprocessorReader.new doc, replacement_content
```

2. You can push a synthetic include (after first emptying the existing reader so we can reuse it)

```ruby
reader.read_lines
reader.push_include replacement_content
```

3. You can restore the lines (after emptying the existing reader so we can reuse it), then force the PreprocessorReader to reprocess all the lines (not recommended, shown for demonstration purposes only)

```ruby
reader.read_lines
reader.restore_lines replacement_content.lines.map {|l| l.rstrip }
reader.instance_variable_set :@look_ahead, 0
reader
```

I'd be willing to consider adding a "reset" (or "reprocess") method to the PreprocessorReader so lines are processed again. We could also add an optional second argument ("preprocess") to `restore_lines` that clears the look_ahead just for the new lines.

cc: @robertpanzer

**Comment by mojavelinux** [comment]
To be honest, the "drop-line" tells me AsciidoctorJ is behaving in a non-compliant way here. It should just treat the include as normal paragraph content (assuming you use `restoreLines` as you described).

**Comment by mojavelinux** [comment]
Take note of the JavaDoc for restoreLines:

```
/**
  * Push multiple lines onto the beginning of the Array of source data.
  *
  * Since this lines were (assumed to be) previously retrieved through the
  * reader, they are marked as seen.
  */
```

**Comment by mojavelinux** [comment]
Unless I'm mistaken, there doesn't seem to be a way to create a PreprocessorReader in AsciidoctorJ. There's an important distinction between a Reader and a PreprocessorReader.

**Comment by mojavelinux** [comment]
Can you show all the lines of your Preprocessor? I'm curious to know if the Preprocessor in AsciidoctorJ is behaving differently than one in Ruby, but I need to make sure we are doing the same thing.

Here's my version in Ruby that attempts to mimic your original report:

```ruby
require 'asciidoctor/extensions'

Asciidoctor::Extensions.register do
  preprocessor do
    process do |doc, reader|
      replacement_content = %(= Document Title
:var: dir/

include::{var}other-file.adoc[]

End of file.)
      reader.read_lines
      reader.unshift_lines replacement_content.lines.map {|l| l.rstrip }
      reader
    end
  end
end
```

I get the following output with no warning:

```
<div class="paragraph">
<p>include::dir/other-file.adoc[]</p>
</div>
<div class="paragraph">
<p>End of file.</p>
</div>
```

Now, the include directive was in the original file, and I didn't clear the reader first, then added new lines onto the reader...then I do get the warning message, but that's because the structure of the file is breaking, not because the preprocessor isn't running. If you use restoreLines, you are putting lines onto the front of the reader, so you better make sure that doesn't break an existing document header. Usually, when using restoreLines, you want the reader to be empty.

**Comment by mojavelinux** [comment]
I'm feeling confident that AsciidoctorJ is not behaving differently...we were just using the Preprocessor differently.

However, we still need to allow a PreprocessorReader to be created. We need a createReader method on PreprocessorReader in AsciidoctorJ, I think.

**Comment by mmews-n4** [comment]
Thanks a lot for your review, @mojavelinux!

I didn't try to use the method `unshift_lines` since it's not available in the Reader or PreprocessorReader, I think. 

**Q1:** In your example above, you get html output with no warning. But why does the html output contain the include macro as text? I would have expected that this include was resolved and the contents of dir/other-file.adoc were inlined, right?

**Let me elaborate on the initial example of this issue**
Lets assume that the file `dir/otherFile.adoc` contains only:
`This is the content of dir/otherFile.adoc! :-)`

When no Preprocessor is used, the complete adoc document looks like:
```
= My Title
:var: dir/
This is the content of dir/otherFile.adoc! :-)
End of file.
```

When the following preprocessor PP1 is used, the complete adoc document looks like:
```
= My Title
:var: dir/
End of file.
because the include line is dropped.
```

**PP1:**
```
public class PP1 extends Preprocessor {
	@Override
	public void process(Document document, PreprocessorReader reader) {
		List<String> lines = reader.readLines();
		reader.restoreLines(lines);
	}
}
```

**Regarding your first response**
**A1:** I'm not quite sure how you mean it. How would I return a new Reader in the `process` method?
**A2:** I tried the `push_include` and I found that normal content lines get duplicated in the resulting text.
**A3:** The method `reader.instance_variable_set` was not available. But this sounds like a solution. (or using an additional parameter 'resetReader')

**Currently, I do the following:**
I set the document variable `var` manually using the statement:
`document.setAttr("var", "dir/", true);`
in the Preprocessor#process method. The result is, that the ruby method `@document.sub_attributes` will find the `var` variable and replaces it with `dir/`. This way, the include line won't get dropped.

**Comment by mojavelinux** [comment]
> I didn't try to use the method unshift_lines

It's the same method as restore_lines (just an alias).

**Comment by mojavelinux** [comment]
> But why does the html output contain the include macro as text? I would have expected that this include was resolved and the contents of dir/other-file.adoc were inlined, right?

Because once we put the lines back onto the reader (assuming the lines we are putting back contain an include macro), it demonstrates that the preprocessor doesn't run again.

**Comment by mojavelinux** [comment]
Hang on. Now I realize what's going on.

It's extremely important to remember that a preprocessor runs before any parsing is done. That means that the document header has not be read yet. And the preprocessor just reads lines for preprocessor directives, nothing else. So it doesn't see the document header. Therefore, a preprocessor has to be pre-populated with all document attributes you want to use in preprocessor lines (such as the include directive).

The way to do this is to parse the document up to the end of the header, collect all those attributes from the header, then pass them to the preprocessor so it can assign them on the document before reading the lines.

This is performed in the following extension from the lab. https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/master/scripts/asciidoc-coalescer.rb#L48-L51

In summary, the preprocessor doesn't see document structure. Therefore, it knows nothing about attributes in the document header.

**Comment by mojavelinux** [comment]
You could also do a lightweight parsing of the header yourself to collect the attributes.

**Comment by mmews-n4** [comment]
Thanks again for your advice!

I understand that the preprocessor is executed before anything else.

Your last comment sounds like the described behaviour is intended which baffles me a little: 
It means that when we use a preprocessor, we suddenly have to care for all include macros that use variables ourselves. If we don't use a preprocessor, we don't have to care for these include macros. _In other words, using a preprocessor has the side effect, that the variable resolution in include macros brakes._ Do you agree?

**Comment by mojavelinux** [comment]
> It means that when we use a preprocessor, we suddenly have to care for all include macros that use variables ourselves. 

That is correct. I've always said that the preprocessor is a very crude extension point. In fact, it has other side effects, such as causing line numbering to be sometimes a little off. It's just something to keep in mind.

I've considered another extension point that provides something like a preparser that lets you modify the lines with some awareness of the document structure, but the way Asciidoctor is implemented currently makes that a bit difficult. In time, perhaps.

So yes, I agree. We need to make it clear in the docs that a preprocessor has some side effects. That might change one day...but for now that's the case.

**Comment by mojavelinux** [comment]
...another option would be to have a body preprocessor. That might actually solve a lot of issues. It would start once the document header is processed. That's possible to implement today, but it requires extra effort. It might be nice to take that extra effort away.

**Comment by mojavelinux** [comment]
(or have a mode of the preprocessor that runs only on the body, after the header has been seen). Interesting idea.

**Comment by mojavelinux** [comment]
Btw, it's not the preprocessor itself that causes this side effect. It's reading the lines using the PreprocessorReader. The preprocessor itself has the option of not touching the reader, in which case there is no side effect.

**Comment by mmews-n4** [comment]
Thanks for your clarifications! I agree that both some info in the documentation and some other kind of processor are very good ideas.

**Comment by mojavelinux** [comment]
I'll follow up with an issue to propose a "body mode" for the preprocessor and also to add some docs. With that, I think we can close this issue. Do you agree?

**Comment by mmews-n4** [comment]
Yes, thanks!
